Crash in `WebEditorClient::clearUndoRedoOperations`

charliewolfe opened the pull request:
#### 2b99f707d57eedadef4838e4806bd6899646c4f0
<pre>
Crash in `WebEditorClient::clearUndoRedoOperations`
<a href="https://bugs.webkit.org/show_bug.cgi?id=298570">https://bugs.webkit.org/show_bug.cgi?id=298570</a>
<a href="https://rdar.apple.com/153031714">rdar://153031714</a>

Reviewed by Pascoe.

Make m_page in WebEditorClient a WeakPtr, since it can be null. WebEditorClient is owned by
WebCore::Page, which may outlive WebKit::WebPage.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::protectedPage const): Deleted.
(WebKit::WebEditorClient::shouldDeleteRange):
(WebKit::WebEditorClient::smartInsertDeleteEnabled):
(WebKit::WebEditorClient::isSelectTrailingWhitespaceEnabled const):
(WebKit::WebEditorClient::shouldBeginEditing):
(WebKit::WebEditorClient::shouldEndEditing):
(WebKit::WebEditorClient::shouldInsertNode):
(WebKit::WebEditorClient::shouldInsertText):
(WebKit::WebEditorClient::shouldChangeSelectedRange):
(WebKit::WebEditorClient::shouldApplyStyle):
(WebKit::WebEditorClient::registerAttachmentIdentifier):
(WebKit::WebEditorClient::registerAttachments):
(WebKit::WebEditorClient::cloneAttachmentData):
(WebKit::WebEditorClient::didInsertAttachmentWithIdentifier):
(WebKit::WebEditorClient::didRemoveAttachmentWithIdentifier):
(WebKit::WebEditorClient::serializedAttachmentDataForIdentifiers):
(WebKit::WebEditorClient::didApplyStyle):
(WebKit::WebEditorClient::didBeginEditing):
(WebKit::WebEditorClient::respondToChangedContents):
(WebKit::WebEditorClient::respondToChangedSelection):
(WebKit::WebEditorClient::didEndUserTriggeredSelectionChanges):
(WebKit::WebEditorClient::updateEditorStateAfterLayoutIfEditabilityChanged):
(WebKit::WebEditorClient::didUpdateComposition):
(WebKit::WebEditorClient::discardedComposition):
(WebKit::WebEditorClient::canceledComposition):
(WebKit::WebEditorClient::didEndEditing):
(WebKit::WebEditorClient::didWriteSelectionToPasteboard):
(WebKit::WebEditorClient::willWriteSelectionToPasteboard):
(WebKit::WebEditorClient::getClientPasteboardData):
(WebKit::WebEditorClient::performTwoStepDrop):
(WebKit::WebEditorClient::registerUndoStep):
(WebKit::WebEditorClient::clearUndoRedoOperations):
(WebKit::WebEditorClient::canUndo const):
(WebKit::WebEditorClient::canRedo const):
(WebKit::WebEditorClient::undo):
(WebKit::WebEditorClient::redo):
(WebKit::WebEditorClient::requestDOMPasteAccess):
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::textFieldDidBeginEditing):
(WebKit::WebEditorClient::textFieldDidEndEditing):
(WebKit::WebEditorClient::textDidChangeInTextField):
(WebKit::WebEditorClient::textDidChangeInTextArea):
(WebKit::WebEditorClient::doTextFieldCommandFromEvent):
(WebKit::WebEditorClient::textWillBeDeletedInTextField):
(WebKit::WebEditorClient::ignoreWordInSpellDocument):
(WebKit::WebEditorClient::learnWord):
(WebKit::WebEditorClient::checkSpellingOfString):
(WebKit::WebEditorClient::checkGrammarOfString):
(WebKit::WebEditorClient::checkTextOfParagraph):
(WebKit::WebEditorClient::updateSpellingUIWithGrammarString):
(WebKit::WebEditorClient::updateSpellingUIWithMisspelledWord):
(WebKit::WebEditorClient::spellingUIIsShowing):
(WebKit::WebEditorClient::getGuessesForWord):
(WebKit::WebEditorClient::requestCheckingOfString):
(WebKit::WebEditorClient::willChangeSelectionForAccessibility):
(WebKit::WebEditorClient::didChangeSelectionForAccessibility):
(WebKit::WebEditorClient::setInputMethodState):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebEditorClientGtk.cpp:
(WebKit::WebEditorClient::handleGtkEditorCommand):
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::hasRichlyEditableSelection):
(WebKit::WebEditorClient::updateStringForFind):
(WebKit::WebEditorClient::overflowScrollPositionChanged):
(WebKit::WebEditorClient::subFrameScrollPositionChanged):
(WebKit::WebEditorClient::shouldAllowSingleClickToChangeSelection const):
(WebKit::WebEditorClient::shouldRevealCurrentSelectionAfterInsertion const):
(WebKit::WebEditorClient::shouldSuppressPasswordEcho const):
(WebKit::WebEditorClient::shouldRemoveDictationAlternativesAfterEditing const):
(WebKit::WebEditorClient::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::didDispatchInputMethodKeydown):
(WebKit::WebEditorClient::uppercaseWord):
(WebKit::WebEditorClient::lowercaseWord):
(WebKit::WebEditorClient::capitalizeWord):
(WebKit::WebEditorClient::substitutionsPanelIsShowing):
(WebKit::WebEditorClient::toggleSmartInsertDelete):
(WebKit::WebEditorClient::isAutomaticQuoteSubstitutionEnabled):
(WebKit::WebEditorClient::toggleAutomaticQuoteSubstitution):
(WebKit::WebEditorClient::toggleAutomaticLinkDetection):
(WebKit::WebEditorClient::isAutomaticDashSubstitutionEnabled):
(WebKit::WebEditorClient::toggleAutomaticDashSubstitution):
(WebKit::WebEditorClient::isAutomaticTextReplacementEnabled):
(WebKit::WebEditorClient::toggleAutomaticTextReplacement):
(WebKit::WebEditorClient::isAutomaticSpellingCorrectionEnabled):

Canonical link: <a href="https://commits.webkit.org/299768@main">https://commits.webkit.org/299768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/437eea7320f38449a1a6da6ffc34dc799251dfbd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39724 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126368 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72104 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/09b6d683-0e5f-4296-af70-7e81f64b9741) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48301 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91149 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60461 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6d0c4cea-1d85-4770-a834-9331adf23d93) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122983 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71704 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/76272952-bb2c-49a8-8893-1212c8cc73db) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31321 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70000 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101766 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25921 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129281 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46951 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99766 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47317 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103816 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99610 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45071 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23092 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43572 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19092 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46813 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52519 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46279 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49628 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47965 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e6d880e057abd78ee840628a160ac182343916be. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50471

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/437eea7320f38449a1a6da6ffc34dc799251dfbd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50471

charliewolfe committed:
Crash in `WebEditorClient::clearUndoRedoOperations`
https://bugs.webkit.org/show_bug.cgi?id=298570
rdar://153031714

Reviewed by Pascoe.

Make m_page in WebEditorClient a WeakPtr, since it can be null. WebEditorClient is owned by
WebCore::Page, which may outlive WebKit::WebPage.

* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::protectedPage const): Deleted.
(WebKit::WebEditorClient::shouldDeleteRange):
(WebKit::WebEditorClient::smartInsertDeleteEnabled):
(WebKit::WebEditorClient::isSelectTrailingWhitespaceEnabled const):
(WebKit::WebEditorClient::shouldBeginEditing):
(WebKit::WebEditorClient::shouldEndEditing):
(WebKit::WebEditorClient::shouldInsertNode):
(WebKit::WebEditorClient::shouldInsertText):
(WebKit::WebEditorClient::shouldChangeSelectedRange):
(WebKit::WebEditorClient::shouldApplyStyle):
(WebKit::WebEditorClient::registerAttachmentIdentifier):
(WebKit::WebEditorClient::registerAttachments):
(WebKit::WebEditorClient::cloneAttachmentData):
(WebKit::WebEditorClient::didInsertAttachmentWithIdentifier):
(WebKit::WebEditorClient::didRemoveAttachmentWithIdentifier):
(WebKit::WebEditorClient::serializedAttachmentDataForIdentifiers):
(WebKit::WebEditorClient::didApplyStyle):
(WebKit::WebEditorClient::didBeginEditing):
(WebKit::WebEditorClient::respondToChangedContents):
(WebKit::WebEditorClient::respondToChangedSelection):
(WebKit::WebEditorClient::didEndUserTriggeredSelectionChanges):
(WebKit::WebEditorClient::updateEditorStateAfterLayoutIfEditabilityChanged):
(WebKit::WebEditorClient::didUpdateComposition):
(WebKit::WebEditorClient::discardedComposition):
(WebKit::WebEditorClient::canceledComposition):
(WebKit::WebEditorClient::didEndEditing):
(WebKit::WebEditorClient::didWriteSelectionToPasteboard):
(WebKit::WebEditorClient::willWriteSelectionToPasteboard):
(WebKit::WebEditorClient::getClientPasteboardData):
(WebKit::WebEditorClient::performTwoStepDrop):
(WebKit::WebEditorClient::registerUndoStep):
(WebKit::WebEditorClient::clearUndoRedoOperations):
(WebKit::WebEditorClient::canUndo const):
(WebKit::WebEditorClient::canRedo const):
(WebKit::WebEditorClient::undo):
(WebKit::WebEditorClient::redo):
(WebKit::WebEditorClient::requestDOMPasteAccess):
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::textFieldDidBeginEditing):
(WebKit::WebEditorClient::textFieldDidEndEditing):
(WebKit::WebEditorClient::textDidChangeInTextField):
(WebKit::WebEditorClient::textDidChangeInTextArea):
(WebKit::WebEditorClient::doTextFieldCommandFromEvent):
(WebKit::WebEditorClient::textWillBeDeletedInTextField):
(WebKit::WebEditorClient::ignoreWordInSpellDocument):
(WebKit::WebEditorClient::learnWord):
(WebKit::WebEditorClient::checkSpellingOfString):
(WebKit::WebEditorClient::checkGrammarOfString):
(WebKit::WebEditorClient::checkTextOfParagraph):
(WebKit::WebEditorClient::updateSpellingUIWithGrammarString):
(WebKit::WebEditorClient::updateSpellingUIWithMisspelledWord):
(WebKit::WebEditorClient::spellingUIIsShowing):
(WebKit::WebEditorClient::getGuessesForWord):
(WebKit::WebEditorClient::requestCheckingOfString):
(WebKit::WebEditorClient::willChangeSelectionForAccessibility):
(WebKit::WebEditorClient::didChangeSelectionForAccessibility):
(WebKit::WebEditorClient::setInputMethodState):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebEditorClientGtk.cpp:
(WebKit::WebEditorClient::handleGtkEditorCommand):
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebEditorClientIOS.mm:
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::hasRichlyEditableSelection):
(WebKit::WebEditorClient::updateStringForFind):
(WebKit::WebEditorClient::overflowScrollPositionChanged):
(WebKit::WebEditorClient::subFrameScrollPositionChanged):
(WebKit::WebEditorClient::shouldAllowSingleClickToChangeSelection const):
(WebKit::WebEditorClient::shouldRevealCurrentSelectionAfterInsertion const):
(WebKit::WebEditorClient::shouldSuppressPasswordEcho const):
(WebKit::WebEditorClient::shouldRemoveDictationAlternativesAfterEditing const):
(WebKit::WebEditorClient::shouldDrawVisuallyContiguousBidiSelection const):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::handleKeyboardEvent):
(WebKit::WebEditorClient::didDispatchInputMethodKeydown):
(WebKit::WebEditorClient::uppercaseWord):
(WebKit::WebEditorClient::lowercaseWord):
(WebKit::WebEditorClient::capitalizeWord):
(WebKit::WebEditorClient::substitutionsPanelIsShowing):
(WebKit::WebEditorClient::toggleSmartInsertDelete):
(WebKit::WebEditorClient::isAutomaticQuoteSubstitutionEnabled):
(WebKit::WebEditorClient::toggleAutomaticQuoteSubstitution):
(WebKit::WebEditorClient::toggleAutomaticLinkDetection):
(WebKit::WebEditorClient::isAutomaticDashSubstitutionEnabled):
(WebKit::WebEditorClient::toggleAutomaticDashSubstitution):
(WebKit::WebEditorClient::isAutomaticTextReplacementEnabled):
(WebKit::WebEditorClient::toggleAutomaticTextReplacement):
(WebKit::WebEditorClient::isAutomaticSpellingCorrectionEnabled):

Canonical link: https://commits.webkit.org/299768@main
